pkg/cri/resource-manager/policy/builtin/memtier/memtier-policy.go

```diff
@@ -345,23 +321,145 @@ func (p *policy) ExportResourceData(c cache.Container) map[string]string {
 	return data
 }
 
+// reallocateResources reallocates the given containers using the given pool hints
+func (p *policy) reallocateResources(containers []cache.Container, pools map[string]string) error {
+	var errors *multierror.Error
+
+	log.Info("reallocating resources...")
+
+	cache.SortContainers(containers)
+
+	for _, c := range containers {
+		p.releasePool(c)
+	}
+	for _, c := range containers {
+		log.Debug("reallocating resources for %s...", c.PrettyName())
+
+		grant, err := p.allocatePool(c, pools[c.GetCacheID()])
```

Comment:
I have a doubt on giving pool hints after changing _allowed_/_reserved_ sets, as the topology tree can be different. Is here any danger that pool hints are cached legacy information that may result in suboptimal workload placement after reconstructing the topology tree?

(Of course it would make sense to minimize moving workloads from a pool to another, especially in short term. But on the other hand, it would sound right to target optimal long term placement in _reallocateResources_.)

Reply:
We are in damage control mode here. We try to avoid/minimize the costly disruption of moving Guaranteed containers across memory controllers/NUMA nodes, yet we (decided that we) need to reallocate them all. Once we release all containers, we sort them primarily by declared importance (QoS class) and secondarily by size (first memory then CPU). So effectively we end up reallocating Guaranteed containers from largest (as in terms of memory request) to smallest first before the rest.

Without using the old pools as NUMA node hints, even if all the Guaranteed containers would nicely fit into their currently allocated pools, there would be zero guarantee that any of them would end up there. The probability for that to happen for any single container is quasi-random, because where the container ended up originally depends on the order of all container allocation and release requests from the very beginning, and the sizes of the containers involved, both in terms of memory and CPU requirements.

So can we end up in sub-optimal overall workload placement after reallocation ? Yes, definitely we can. But that is not our primary concern here. That can easily happen just by a normal sequence of allocations and releases without any reconfiguration in between. We primarily try to avoid disruption of Guaranteed containers, if/when that is (easily) avoidable. 

With rebalancing, we might want to support opting for overall optimal workload placement instead of avoiding disruption. However, IMO there needs to be an explicit indication for that (e.g. as a preference among the configuration).

Reply:
Ok, not disrupting sounds good default policy.

Reply:
I think the logic here is fine.